crates/bevy_app/src/app.rs

```diff
@@ -359,7 +359,22 @@ impl App {
         // These are different for loops to avoid conflicting access to self
         for variant in S::variants() {
             self.add_schedule(OnEnter(variant.clone()), Schedule::new());
-            self.add_schedule(OnExit(variant), Schedule::new());
+            self.add_schedule(OnExit(variant.clone()), Schedule::new());
+
+            for second in S::variants() {
+                if second == variant {
+                    // skip adding OnTransition for A -> A
+                    continue;
+                }
+
+                self.add_schedule(
+                    OnTransition {
+                        from: variant.clone(),
+                        to: second,
+                    },
+                    Schedule::new(),
+                );
```

Comment:
This eagerly registers schedules that scales quadratically with the number of variants in a state set. Not all of them are going to be used, and will still use memory. Is there a way to do this lazily?

Reply:
If we merge https://github.com/bevyengine/bevy/pull/7911 first we can actually remove all of the schedule registration logic from this method. Which I think I prefer.

Reply:
I would prefer this as well. We wait for #7911 then.